types/eventbus/event_bus.go

```diff
@@ -0,0 +1,241 @@
+package eventbus
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/log"
+	tmpubsub "github.com/tendermint/tendermint/libs/pubsub"
+	"github.com/tendermint/tendermint/libs/service"
+	"github.com/tendermint/tendermint/types"
+)
+
+type Subscription interface {
+	ID() string
+	Next(context.Context) (tmpubsub.Message, error)
+}
+
+// EventBus is a common bus for all events going through the system. All calls
+// are proxied to underlying pubsub server. All events must be published using
+// EventBus to ensure correct data types.
+type EventBus struct {
+	service.BaseService
+	pubsub *tmpubsub.Server
+}
+
+// NewDefault returns a new event bus with default options.
+func NewDefault() *EventBus {
+	pubsub := tmpubsub.NewServer(tmpubsub.BufferCapacity(0))
+	b := &EventBus{pubsub: pubsub}
+	b.BaseService = *service.NewBaseService(nil, "EventBus", b)
+	return b
+}
+
+func (b *EventBus) SetLogger(l log.Logger) {
+	b.BaseService.SetLogger(l)
+	b.pubsub.SetLogger(l.With("module", "pubsub"))
+}
+
+func (b *EventBus) OnStart() error {
+	return b.pubsub.Start()
+}
+
+func (b *EventBus) OnStop() {
+	if err := b.pubsub.Stop(); err != nil {
+		b.pubsub.Logger.Error("error trying to stop eventBus", "error", err)
+	}
+}
+
+func (b *EventBus) NumClients() int {
+	return b.pubsub.NumClients()
+}
+
+func (b *EventBus) NumClientSubscriptions(clientID string) int {
+	return b.pubsub.NumClientSubscriptions(clientID)
+}
+
+// Deprecated: Use SubscribeWithArgs instead.
+func (b *EventBus) Subscribe(ctx context.Context,
+	clientID string, query tmpubsub.Query, capacities ...int) (Subscription, error) {
+
+	return b.pubsub.Subscribe(ctx, clientID, query, capacities...)
+}
+
+func (b *EventBus) SubscribeWithArgs(ctx context.Context, args tmpubsub.SubscribeArgs) (Subscription, error) {
+	return b.pubsub.SubscribeWithArgs(ctx, args)
+}
+
+func (b *EventBus) Unsubscribe(ctx context.Context, args tmpubsub.UnsubscribeArgs) error {
+	return b.pubsub.Unsubscribe(ctx, args)
+}
+
+func (b *EventBus) UnsubscribeAll(ctx context.Context, subscriber string) error {
+	return b.pubsub.UnsubscribeAll(ctx, subscriber)
+}
+
+func (b *EventBus) Observe(ctx context.Context, observe func(tmpubsub.Message) error, queries ...tmpubsub.Query) error {
+	return b.pubsub.Observe(ctx, observe, queries...)
+}
+
+func (b *EventBus) Publish(eventValue string, eventData types.TMEventData) error {
+	// no explicit deadline for publishing events
```

Comment:
what if the service stops should we get a context object that hangs off of that?

Reply:
There is no context tied to the lifecycle of a Service implementation. Are you suggesting adding the parameter pre-emptively? I'm fine with that, I just didn't do it because it wasn't adding any value yet.